src/node_dns.cc

```diff
@@ -0,0 +1,471 @@
+  transactions_[tid].Reset(env()->isolate(), p);
+  return p->GetPromise();
+}
+
+void DNSWrap::ResolveTransaction(getdns_transaction_t tid, Local<Value> v) {
+  auto it = transactions_.find(tid);
+  it->second.Get(env()->isolate())->Resolve(env()->context(), v).ToChecked();
+  transactions_.erase(it);
+}
+
+void DNSWrap::RejectTransaction(getdns_transaction_t tid, Local<Value> v) {
+  auto it = transactions_.find(tid);
+  it->second.Get(env()->isolate())->Reject(env()->context(), v).ToChecked();
+  transactions_.erase(it);
+}
+
+// static
+void DNSWrap::GetAddresses(const FunctionCallbackInfo<Value>& args) {
+  DNSWrap* dns;
+  ASSIGN_OR_RETURN_UNWRAP(&dns, args.This());
+  Environment* env = dns->env();
+  CHECK(args[0]->IsString());
+  String::Utf8Value name(env->isolate(), args[0]);
+  getdns_dict* extension = nullptr;
+  getdns_transaction_t tid;
+  GETDNS(address,
+         env->isolate(),
+         dns->context_,
+         *name,
+         extension,
+         dns,
+         &tid,
+         Callback);
+  args.GetReturnValue().Set(dns->RegisterTransaction(tid));
+}
+
+// static
+void DNSWrap::GetServices(const FunctionCallbackInfo<Value>& args) {
+  DNSWrap* dns;
+  ASSIGN_OR_RETURN_UNWRAP(&dns, args.This());
+  Environment* env = dns->env();
+  CHECK(args[0]->IsString());
+  String::Utf8Value name(env->isolate(), args[0]);
+  getdns_dict* extension = nullptr;
+  getdns_transaction_t tid;
+  GETDNS(service,
+         env->isolate(),
+         dns->context_,
+         *name,
+         extension,
+         dns,
+         &tid,
+         Callback);
+  args.GetReturnValue().Set(dns->RegisterTransaction(tid));
+}
+
+static bool ConvertAddress(Isolate* isolate,
+                           const char* address,
+                           getdns_dict* dict) {
+  char address_buffer[sizeof(struct in6_addr)];
+  size_t length;
+  int family;
+  if (uv_inet_pton(AF_INET, address, &address_buffer) == 0) {
+    length = sizeof(struct in_addr);
+    family = AF_INET;
+  } else if (uv_inet_pton(AF_INET6, address, &address_buffer) == 0) {
+    length = sizeof(struct in6_addr);
+    family = AF_INET6;
+  } else {
+    THROW_ERR_DNS_ERROR(isolate, "Invalid address");
+    return false;
+  }
+
+  getdns_bindata address_data{length,
+                              reinterpret_cast<uint8_t*>(address_buffer)};
+  GETDNSV(dict_util_set_string,
+          isolate,
+          dict,
+          "address_type",
+          family == AF_INET ? "IPv4" : "IPv6");
+  GETDNSV(dict_set_bindata, isolate, dict, "address_data", &address_data);
+
+  return true;
+}
+
+// static
+void DNSWrap::GetHostnames(const FunctionCallbackInfo<Value>& args) {
+  DNSWrap* dns;
+  ASSIGN_OR_RETURN_UNWRAP(&dns, args.This());
+  Environment* env = dns->env();
+
+  CHECK(args[0]->IsString());
+  String::Utf8Value name(env->isolate(), args[0]);
+  getdns_dict* address = getdns_dict_create_with_context(dns->context_);
+  auto cleanup = OnScopeLeave([&]() {
+    getdns_dict_destroy(address);
+  });
```

Comment:
a `using DnsDictPointer = DeleteFnPtr<getdns_dict, getdns_dict_destroy>` would likely be useful here